[196] Add support for MSSQL

RobertLD opened the pull request:
Add framework for SQLServer support.

Currently the specification differs from what is expected by the SODA-CLI so this will have to be updated to match before the added tests will pass 

https://docs.soda.io/soda/connect-mssql.html


Related to: #196

simonharrer commented on tests/fixtures/sqlserver/datacontract.yaml:
```diff
@@ -0,0 +1,28 @@
+        type: varchar
+        required: true
+        unique: true
+        regex: "[A-Za-z]{2}-\\d{3}-[A-Za-z]{2}$"
```
should probably be pattern

RobertLD replied:
Hey Simon, how do you mean?

simonharrer replied:
regex is not a known field in the data contract spec, should be named pattern, I think.

simonharrer replied:
a then it is wrong there as well. Good catch! :-)

simonharrer commented on datacontract/model/data_contract_specification.py:
```diff
@@ -31,6 +31,10 @@ class Server(pyd.BaseModel):
+    trusted_connection: str = None
+    trust_server_certificate: str = None
+    encrypt_connection: str = None
+    driver: str = None # Use ENV variable
```
do you need those fields? could we get by without them?

RobertLD replied:
The default provided values are not entirely suitable for a large majority
of connections. Users that are using Kerberos or active directory, for
example require a trusted connection and will not provide a username or
password.

RobertLD replied:
> do you need those fields? could we get by without them?

I am open to alternatives if you'd rather have them be defined as environmental variables, for example

jochenchrist replied:
I'd go for environment variables and sensible defaults, as they probably would not be part of the datacontract.yaml

simonharrer replied:
Hm, good question. I am no export of sqlserver, so I don't really know which way is better.

For me, env variables should be used for secrets or things that differ from user to user.

If, however, you need those fields in order to know how to connect to the data, and it can only be done knowing those infos, then we should add them to the spec. 

Does this help? What would you do with this guideline?

PS: I just noticed that the three fields are of type boolean, but represented as strings in the pydantic data contract spec model.

RobertLD replied:
@simonharrer I wanted to match the current code style as much as possible and use string representation as that seemed to be the default.

The only one absolutely required without any reasonable defaults is MSSQL driver. The others can be made as env variables

RobertLD commented:
The default provided values are not entirely suitable for a large majority
of connections. Users that are using Kerberos or active directory, for
example require a trusted connection and will not provide a username or
password.

On Wed, May 15, 2024, 2:59 PM Dr. Simon Harrer ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In datacontract/model/data_contract_specification.py
> <https://github.com/datacontract/datacontract-cli/pull/204#discussion_r1602116891>
> :
>
> > @@ -31,6 +31,10 @@ class Server(pyd.BaseModel):
>      token: str = None  # Use ENV variable
>      dataProductId: str = None
>      outputPortId: str = None
> +    trusted_connection: str = None
> +    trust_server_certificate: str = None
> +    encrypt_connection: str = None
> +    driver: str = None # Use ENV variable
>
> do you need those fields? could we get by without them?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/datacontract/datacontract-cli/pull/204#pullrequestreview-2058791087>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHGZJYRTUYCMM4TPFKADGJ3ZCOWCTAVCNFSM6AAAAABHYTVY52VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDANJYG44TCMBYG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

RobertLD commented:
It certainly can be. I'll make that change.

On Wed, May 15, 2024, 2:57 PM Dr. Simon Harrer ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pyproject.toml
> <https://github.com/datacontract/datacontract-cli/pull/204#discussion_r1602114655>
> :
>
> > @@ -40,7 +40,8 @@ dependencies = [
>    "opentelemetry-exporter-otlp-proto-http~=1.16.0",
>    "deltalake~=0.17.0",
>    "boto3>=1.34.41,<1.34.99",
> -  "botocore>=1.34.41,<1.34.99"
> +  "botocore>=1.34.41,<1.34.99",
> +  "pymssql==2.3.0"
>
> can this be a dev dependency?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/datacontract/datacontract-cli/pull/204#pullrequestreview-2058787411>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHGZJYQYYU2SIR2FO3TOKADZCOV25AVCNFSM6AAAAABHYTVY52VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDANJYG44DONBRGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

RobertLD commented on pyproject.toml:
```diff
@@ -40,7 +40,8 @@ dependencies = [
   "boto3>=1.34.41,<1.34.99",
-  "botocore>=1.34.41,<1.34.99"
+  "botocore>=1.34.41,<1.34.99",
+  "pymssql==2.3.0"
```
It certainly can be. I'll make that change.

jochenchrist commented:
@RobertLD There was an error in the Data Contract JSON Schema (https://datacontract.com/datacontract.schema.json). I fixed the list of required fields there.

RobertLD commented:
> @RobertLD There was an error in the Data Contract JSON Schema (https://datacontract.com/datacontract.schema.json). I fixed the list of required fields there.

Sorry to bother you again but there appears to be one more issue with the schema. SQLServer types aren't supported in the field types yet. There are a whole bunch of them, but for now datetime and datetime2 are missing.

![image](https://github.com/datacontract/datacontract-cli/assets/30250210/ac002c87-2806-4690-bf76-dde9d334c845)

jochenchrist commented:
> Sorry to bother you again but there appears to be one more issue with the schema. SQLServer types aren't supported in the field types yet. There are a whole bunch of them, but for now datetime and datetime2 are missing.

The `field.type` represents a logical type, and it is a finite list of types. https://datacontract.com/#data-types
You'd need to define a mapping to / from these types.

In addition, you can specify the physical MS SQL Server Types in a config block, like this: https://datacontract.com/#config-object

```
models:
  my_table:
    fields:
      my_field_1:
        description: Example for a timestamp
        type: timestamp
        config:
          sqlserverType: DATETIME2
```

This config object is quite new, if you have any improvement ideas, I'd be happy to hear them.

RobertLD commented:
> type: timestamp_ntz
>         config:
>           sqlserverType: DATETIME2

I updated the datacontract to use the config object and it doesn't appear to be actually overwriting the type to check against.

![image](https://github.com/datacontract/datacontract-cli/assets/30250210/1d096b3f-1059-4c4f-97e4-3ae7d60439d2)

![image](https://github.com/datacontract/datacontract-cli/assets/30250210/34e57222-64f0-4ab8-b412-18ddd27b5cbc)

![image](https://github.com/datacontract/datacontract-cli/assets/30250210/e54e7098-71e8-4d37-a582-1fdf196b9455)

We're in the end-zone now; this is the last element before I formalize the work here and rebase to something cleaner. Thank you a lot for the support in doing the work!

simonharrer commented:
Thanks already for your dedication.

You would need something like this for the type mapping
https://github.com/datacontract/datacontract-cli/commit/b6bd8fbec07a0531698e6cf061a19c3c72bbe041

And we need a type mapping for sqlserver anyway in datacontract/export/sql_type_converter.py

RobertLD commented:
> Thanks already for your dedication.
> 
> You would need something like this for the type mapping [b6bd8fb](https://github.com/datacontract/datacontract-cli/commit/b6bd8fbec07a0531698e6cf061a19c3c72bbe041)
> 
> And we need a type mapping for sqlserver anyway in datacontract/export/sql_type_converter.py

Added the serializer for sqlserver data-types and added the correct driver to the docker file. All of the functionality ought to be there now.


I've also rebased ontop of main so it should be a bit easier to review now

simonharrer commented on datacontract/engines/soda/connections/sqlserver.py:
```diff
@@ -0,0 +1,41 @@
+            "schema": server.schema_,
+            "trusted_connection": os.getenv("DATACONTRACT_TRUSTED_CONNECTION", False),
+            "trust_server_certificate": os.getenv("DATACONTRACT_TRUST_SERVER_CERTIFICATE", False),
+            "encrypt":  os.getenv("DATACONTRACT_ENCRYPTED_CONNECTION", True),
```
I'd prefix the other settings to DATACONTRACT_SQLSERVER_ as well

RobertLD replied:
Fair enough; I'll make them more consistent

simonharrer commented on Dockerfile:
```diff
@@ -22,7 +22,7 @@ RUN python -c "import duckdb; duckdb.connect().sql(\"INSTALL httpfs\");"
 FROM ubuntu:22.04 AS runner-image
 
-RUN apt-get update && apt-get install --no-install-recommends -y python3.11 python3.11-venv && \
+RUN apt-get update && apt-get install --no-install-recommends -y python3.11 python3.11-venv msodbcsql18 && \
```
Why is this necessary? Can you share a little bit on that?

simonharrer replied:
We do not use the Dockerfile for test execution. We run the tests with pytest locally, and we use github ci to execute them on CI. Testcontainers is only used to provision databases so we test against the actual thing. 

We should install that driver in the github workflow file instead. And we should disable the test when not in CI as it would otherwise force everybody to install those drivers on their machines to male the tests run.

RobertLD replied:
These two statements seem to be conflicting. If you want to execute that test and have it succeed it requires the driver. As for the CI I can add it there as well, but as it stands I'm unsure what the consensus is for having it in the Dockerfile as well

simonharrer replied:
I think both comments are valid. We need the lib in the Dockerfile for `datacontract test` command, but we also need the lib in the CI pipeline so we can run `pytest` in GitHub actions as well as run the test locally.

RobertLD replied:
Added to both areas

RobertLD replied:
Updated to skip the test outside of CI/CD and likewise removed from dockerfile

simonharrer reviewed (changes requested):
Nice work. Only two minor things.

Additionally, please add an entry to CHANGELOG.md what you have done and make sure to add the server type in the README.md

Future work could be to support for `datacontract export --format sql` for sqlserver along with a test and make sure that the types are correctly set during `datacontract import --format sql`.

RobertLD commented:
> Nice work. Only two minor things.
> 
> Additionally, please add an entry to CHANGELOG.md what you have done and make sure to add the server type in the README.md
> 
> Future work could be to support for `datacontract export --format sql` for sqlserver along with a test and make sure that the types are correctly set during `datacontract import --format sql`.


Made the requested changes to the changelog and readme. As for the future work, I can put together two issues for those to stop this issue from expanding to gigantic proportions. I'll start working on those next; if you're alright with splitting them out into separate issues

simonharrer commented:
Sure, feel free to create those issues.

RobertLD commented:
> My last issue is, that in order to run the tests, I now have to install the ODBC driver on my machine.
> 
> Instead, I'd like that this sqlserver test is ignored by default, and only executed in the CI/CD pipeline. Can you implement that? Otherwise, it is good to go. Thanks for your dedication here!

Oh course! I'll implement these changes

RobertLD commented:
> My last issue is, that in order to run the tests, I now have to install the ODBC driver on my machine.
> 
> Instead, I'd like that this sqlserver test is ignored by default, and only executed in the CI/CD pipeline. Can you implement that? Otherwise, it is good to go. Thanks for your dedication here!

Changes made; easier than anticipated haha

simonharrer commented on .vscode/settings.json:
can you please remove this file?

RobertLD replied:
Accidental inclusion, removed

simonharrer commented on Dockerfile:
```diff
@@ -22,7 +22,7 @@ RUN python -c "import duckdb; duckdb.connect().sql(\"INSTALL httpfs\");"
 FROM ubuntu:22.04 AS runner-image
 
-RUN apt-get update && apt-get install --no-install-recommends -y python3.11 python3.11-venv msodbcsql18 && \
+RUN apt-get update && apt-get install --no-install-recommends -y python3.11 python3.11-venv && \
```
I'm confused. Why did you remove this here?

RobertLD replied:
I guess I misunderstood you; I thought we didn't want it included in the dockerfile but only in the git workflow?

simonharrer replied:
I'm fine with it being in the Dockerfile as we'll ship the CLI as a docker container, and it should come prepacked with necessary software packages that it works out of the box.

simonharrer commented on tests/test_test_sqlserver.py:
```diff
@@ -23,7 +24,7 @@ def remove_container():
     request.addfinalizer(remove_container)
 
-
+@pytest.mark.skipif(not os.getenv('CI'), reason="Skipping test outside CI/CD environment")
```
Ah, simple enough. Thanks. :-)

RobertLD commented:
I'll add it back in a few hours, then. My mistake.

On Wed, May 22, 2024, 4:55 PM Dr. Simon Harrer ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In Dockerfile
> <https://github.com/datacontract/datacontract-cli/pull/204#discussion_r1610637288>
> :
>
> > @@ -22,7 +22,7 @@ RUN python -c "import duckdb; duckdb.connect().sql(\"INSTALL httpfs\");"
>
>  FROM ubuntu:22.04 AS runner-image
>
> -RUN apt-get update && apt-get install --no-install-recommends -y python3.11 python3.11-venv msodbcsql18 && \
> +RUN apt-get update && apt-get install --no-install-recommends -y python3.11 python3.11-venv && \
>
> I'm fine with it being in the Dockerfile as we'll ship the CLI as a docker
> container, and it should come prepacked with necessary software packages
> that it works out of the box.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/datacontract/datacontract-cli/pull/204#discussion_r1610637288>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHGZJYRWO5Q44NT4WZJHROLZDUA47AVCNFSM6AAAAABHYTVY52VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDANZSGM2DQNRTGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

simonharrer commented:
Thanks for your dedication to bring this PR through! :-)

RobertLD commented:
Of course, thank you for the assistance in getting it done. I'll follow up
with those smaller issues shortly and hopefully knock those out soon.

On Fri, May 24, 2024, 6:47 AM Dr. Simon Harrer ***@***.***>
wrote:

> Merged #204 <https://github.com/datacontract/datacontract-cli/pull/204>
> into main.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/datacontract/datacontract-cli/pull/204#event-12923254020>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHGZJYV4CWWKRTECY6DJTDDZD4LFBAVCNFSM6AAAAABHYTVY52VHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJSHEZDGMRVGQYDEMA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>
